lib/puppet/pops/loader/module_loaders.rb

```diff
@@ -424,8 +424,13 @@ def existing_path(effective_path)
     end
 
     def existing_paths(effective_path)
-      # Select all paths starting with effective_path but reject any path that continues into a subdirectory
-      @path_index.select { |path| path.start_with?(effective_path) }.reject { |path| path[effective_path.size..-1].include?('/')}
+      dirname = File.dirname(effective_path)
+      basename = File.basename(effective_path)
+      # Select all paths matching `effective_path` up until an optional file extension
+      @path_index.select do |path|
+        File.basename(path, '.*') == basename &&
```

Comment:
This won't work for paths like `foo.tar.gz` since the basename of that is `foo.tar` which in turn is an illegal task name (tasks cannot contain dots). I suggest using the regexp used here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/loader/loader_paths.rb#L255

Reply:
That's a tricky case. I think that `foo.tar.gz` *should* be an invalid task file. The question is whether the filename is everything until the *first* dot, or whether the extension is everything after the *last* dot. Standard practice seems to be the latter.

Since a legitimate task file is extremely unlikely to actually end in `.tar.gz` (as that's not generally executable...), I wonder whether there's a more realistic case of this problem that we might run into. If not, I would prefer to be strict.

Reply:
That's one way of looking at it. On the other hand, a task name cannot contain a dot so why not resolve this in a consistent manner? If we want to be more strict, then we need to change the referenced regexp as well. We can go either way but the approach taken should be consistently implemented.

Reply:
Stripping only the last extension is the behavior used by orchestrator in PE today (per [this code](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/module/task.rb#L86-L88), so I'll update the code you referenced for consistency.

Reply:
While trying to write a failing test, it looks like that code didn't actually fully support `foo.tar.gz` anyway. The `discover` method rules out that file as invalid, using [the same logic](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/loader/loader_paths.rb#L278-L284) as I'm using here.

Reply:
Glad we got that one sorted out (note to self, never complain about something you haven't written tests for).